File: mesh/generic/li1Radio.py
from mesh.generic.radio import Radio
from mesh.generic.checksum import calc8bitFletcherChecksum, compareChecksum
from mesh.generic.li1RadioCmds import Li1RadioCmds, Li1RadioPayloadCmds
from struct import pack, unpack
from math import ceil

Li1HeaderLength = 8
Li1SyncBytes = b'He'
lenSyncBytes = len(Li1SyncBytes)
checksumLen = 2
Li1MaxPayload = 255
class Li1Radio(Radio):
    
    def __init__(self, serial, config):
        Radio.__init__(self, serial, config)
        self.cmdRxBuffer = bytearray()

    def createCommand(self, cmd):
        """Send command to Li-1 Radio."""
        self.createHeader(cmd) # header
        self.createPayload(cmd) # payload
        return cmd['msgBytes']

    def sendMsg(self, msgBytes):
        """Send messages to radio."""
        if len(msgBytes) > 0:
            numMsgs = ceil(len(msgBytes)/Li1MaxPayload) # calculate how many radio messages required
            for i in range(numMsgs):
                # Split into multiple messages
                if len(msgBytes) >= (i+1)*Li1MaxPayload: # get message end
                    msgEnd = (i+1)*Li1MaxPayload
                else:
                    msgEnd = len(msgBytes)
                payload = msgBytes[i*Li1MaxPayload:msgEnd]

                msg = self.createMsg(payload)
                self.serial.write(msg)
    
    def createMsg(self, rawMsg):
        """Create message to send to Li-1 radio."""
        # Create header
        msg = {'commandType': Li1RadioCmds['Transmit'], 'payloadSize': len(rawMsg), 'payload': rawMsg, 'msgBytes': bytearray()} # wrap message in command container
        #msg['payload'] = b'\xc0'*100
        #msg['payloadSize'] = len(msg['payload'])
        self.createCommand(msg)
        
        return msg['msgBytes']

    def createHeader(self, msg):
        """Create message header for Li-1 radio message."""
        # Create header
        #headerBytes = bytearray(Li1HeaderLength)
        headerBytes = Li1SyncBytes # sync bytes
        headerBytes += pack('>H', msg['commandType']) # command type, hex unique command ID sent as unsigned short
        headerBytes += pack('>H', msg['payloadSize']) # payload size, big endian unsigned short integer

        # Create and append checksum
        headerBytes += pack('BB', *calc8bitFletcherChecksum(headerBytes[2:]))
         
        msg['msgBytes'] += headerBytes

    def createPayload(self, msg):
        """Create payload for Li-1 message packet."""
        # Create and append payload checksum
        if 'payload' in msg:
            payloadBytes = msg['payload'] + pack('BB', *calc8bitFletcherChecksum(msg['msgBytes'][lenSyncBytes:] + msg['payload']))
            msg['msgBytes'] += payloadBytes
        else: # no payload
            return
    

    def processRxBytes(self, serBytes, bufferFlag):
        # Search received bytes for commands
        self.cmdRxBuffer += serBytes
        msgEnd = 0
        while msgEnd < len(self.cmdRxBuffer): # Continue searching until end of received bytes
            msgEnd, cmd = self.parseCommand(self.cmdRxBuffer)
            # Sort commands
            if cmd:
                print("Command rx buffer:", self.cmdRxBuffer)
                print(cmd)
                self.cmdRxBuffer = self.cmdRxBuffer[msgEnd:] # clear out parsed bytes
                # Buffer command
                self.cmdBuffer.append(cmd)

                # Check for data
                if cmd['cmdType'] == Li1RadioCmds['ReceivedData'] and cmd['payload'] != None: # Valid data message
                    # Put data in rxBuffer
                    print("Valid data message received.")
                    self.bufferRxMsg(cmd['payload'], bufferFlag)
                else:  # Other command type
                    print(cmd)
                    pass
        if len(self.cmdRxBuffer) > 1000:
            self.cmdRxBuffer = bytearray()

    def parseCmdHeader(self, cmd, serBytes):
        # Parse header
        headerBytes = serBytes[0:Li1HeaderLength] 
        cmdInfo = headerBytes[lenSyncBytes:-checksumLen]
        checksumBytes = headerBytes[-checksumLen:]
                    
        # Check header checksum
        if (compareChecksum(cmdInfo, checksumBytes)):
            #print("Header checksum matches")
            cmd['header'] = cmdInfo
            cmd['rawHeader'] = headerBytes[lenSyncBytes:]
            # Decode command type
            cmd['cmdType'] = unpack('>H', cmd['header'][0:2])[0]
            return True # header found
        else:
            print("Header checksum failed")
            return False # header not found
        

    def parseCmdPayload(self, payloadSize, serBytes, headerBytes):
        # Parse message payload (if present)
        if len(serBytes) >= (payloadSize + checksumLen): # entire message received
            msgEnd = payloadSize + checksumLen
            # Check payload checksum
            payloadBytes = serBytes[0:payloadSize]
            payloadChecksumBytes = serBytes[payloadSize:payloadSize+checksumLen]

            if (compareChecksum(headerBytes+payloadBytes, payloadChecksumBytes)):
                #print("Payload checksum matches")
                #return msgEnd, payloadBytes
                return msgEnd, self.parseAX25Msg(payloadBytes)
            else: # invalid checksum
                print("Invalid payload checksum")
                return msgEnd, None
                
        else: # incomplete command
                #print("Incomplete payload")
                return len(serBytes), None

    def parseAX25Msg(self, msgBytes):
        """Parse AX.25 formatted message."""
        # For now, just discard everything but original message payload."""        
        return msgBytes[16:-2] 

    def parseCommand(self, serBytes):
        """Search raw received bytes for commands."""
    
        cmd = {'header': None, 'payload': None}

        for i in range(len(serBytes) - (lenSyncBytes-1)):
            # Search serial bytes for sync characters
            if (serBytes[i:i+lenSyncBytes] == Li1SyncBytes): # sync bytes found
                #print("Message Found")
                msgStart = i
               
                if (len(serBytes[msgStart:]) >= Li1HeaderLength): # entire header received
                    # Parse command header
                    headerFound = self.parseCmdHeader(cmd, serBytes[msgStart:])
                else: # incomplete command
                    return len(serBytes), None

                # Update buffer position
                msgEnd = msgStart + Li1HeaderLength

                if headerFound:
                    # Check if payload present
                    if cmd['cmdType'] in Li1RadioPayloadCmds.values():
                        payloadSize = unpack('>H', cmd['header'][2:])[0]
                        if payloadSize == 0: # No payload
                            return msgEnd, cmd # header only command
                        elif payloadSize == 65535: # receive error
                            return msgEnd, cmd
                        else: # payload present
                            if len(serBytes) >= (Li1HeaderLength + payloadSize + checksumLen): # entire message received
                                payloadEnd, cmd['payload'] = self.parseCmdPayload(payloadSize, serBytes[msgStart+Li1HeaderLength:], cmd['rawHeader'])
                                msgEnd += payloadEnd # update buffer position
                                if cmd['payload']:
                                    return msgEnd, cmd
                                else:
                                    return msgEnd, None # invalid payload
                            else: # incomplete command
                                return len(serBytes), None
                        
                    else: # no payload
                        return msgEnd, cmd # header only command
            
                else: # invalid header
                    return msgEnd, None

        return len(serBytes), None

    def sendCommand(self, cmd):
        """Issue command to Li-1 radio."""
        
        # Send no-op command
        noOpCmd = {'commandType': Li1RadioCmds['NoOp'], 'payloadSize': 0, 'msgBytes': bytearray()} # wrap message in command container
        self.createHeader(noOpCmd) # header
        self.createPayload(noOpCmd) # payload
        self.serial.write(noOpCmd['msgBytes'])

File: cpp/comm/li1Radio.cpp
#include "comm/li1Radio.hpp"
#include "comm/checksum.hpp"
#include "comm/utilities.hpp"
#include <math.h>

using std::vector;



namespace comm {
    
    // Static members
    vector<uint8_t> Li1Radio::syncBytes = {72, 101}; // 'He'
    unsigned int Li1Radio::lenSyncBytes = Li1Radio::syncBytes.size();
    unsigned int Li1Radio::headerLength = 8;
    unsigned int Li1Radio::checksumLen = 2;
    unsigned int Li1Radio::maxPayload = 255;

    Li1Radio::Li1Radio() : 
        Radio(),
        cmdBuffer(CommBuffer<Li1Cmd>(5)),
        bigendian(util::isbigendian())
    {
        cmdRxBuffer.reserve(config.cmdRxBufferSize);
    }

    Li1Radio::Li1Radio(serial::Serial * serialIn, RadioConfig & configIn) :
        Radio(serialIn, configIn),
        cmdBuffer(CommBuffer<Li1Cmd>(5)),
        bigendian(util::isbigendian())
    {
        cmdRxBuffer.reserve(config.cmdRxBufferSize);
    }

    void Li1Radio::createHeader(Li1Cmd & cmd) {
        // Start with sync bytes
        vector<uint8_t> headerBytes = Li1Radio::syncBytes;
        if (bigendian == true) {
            // Add command type (big-endian)
            headerBytes.push_back((uint8_t)(cmd.type & 0x00FF));
            headerBytes.push_back((uint8_t)((cmd.type & 0xFF00) >> 8));
            // Add payload size (big-endian)
            headerBytes.push_back((uint8_t)(cmd.payload.size() & 0x00FF));
            headerBytes.push_back((uint8_t)((cmd.payload.size() & 0xFF00) >> 8));
        }
        else {
            // Add command type (big-endian)
            headerBytes.push_back((uint8_t)((cmd.type & 0xFF00) >> 8));
            headerBytes.push_back((uint8_t)(cmd.type & 0x00FF));
            // Add payload size (big-endian)
            headerBytes.push_back((uint8_t)((cmd.payload.size() & 0xFF00) >> 8));
            headerBytes.push_back((uint8_t)(cmd.payload.size() & 0x00FF));
        }   
        // Append checksum
        vector<uint8_t> checksum = calculate8bitFletcherChecksum(vector<uint8_t>(headerBytes.begin()+Li1Radio::lenSyncBytes, headerBytes.end()));
        headerBytes.insert(headerBytes.end(), checksum.begin(), checksum.end());
        // Add to message bytes
        cmd.msgBytes.insert(cmd.msgBytes.end(), headerBytes.begin(), headerBytes.end());
    }
    
    void Li1Radio::createPayload(Li1Cmd & cmd) {
        if (cmd.payload.size() > 0) {
            // Append payload bytes
            cmd.msgBytes.insert(cmd.msgBytes.end(), cmd.payload.begin(), cmd.payload.end());
            // Append checksum
            vector<uint8_t> checksum = calculate8bitFletcherChecksum(vector<uint8_t>(cmd.msgBytes.begin()+Li1Radio::lenSyncBytes, cmd.msgBytes.end()));
            cmd.msgBytes.insert(cmd.msgBytes.end(), checksum.begin(), checksum.end());
        }
    }
    
    void Li1Radio::createCommand(Li1Cmd & cmd) {
        // Create command elements
        createHeader(cmd);
        createPayload(cmd);
    }
    
    vector<uint8_t> Li1Radio::createMsg(vector<uint8_t> msgBytes) {
        // Create command
        Li1Cmd cmd(LI1_TRANSMIT, msgBytes);
        createCommand(cmd);

        // Return raw command bytes
        return cmd.msgBytes;
    }
    
    bool Li1Radio::parseCmdHeader(Li1Cmd & cmd, const vector<uint8_t> serBytes) {
        // Header components
        vector<uint8_t> headerBytes = vector<uint8_t>(serBytes.begin(), serBytes.begin() + Li1Radio::headerLength);    
        vector<uint8_t> cmdInfo = vector<uint8_t>(headerBytes.begin() + Li1Radio::lenSyncBytes, headerBytes.end() - Li1Radio::checksumLen);
        vector<uint8_t> checksumBytes = vector<uint8_t>(headerBytes.end() - Li1Radio::checksumLen, headerBytes.end());       
        // Check header checksum
        if (compareChecksum(cmdInfo, checksumBytes) == true) { // valid checksum
            cmd.header = cmdInfo;
            cmd.rawHeader = vector<uint8_t>(headerBytes.begin() + Li1Radio::lenSyncBytes, headerBytes.end());
            cmd.type = ((uint16_t)cmd.header[0] << 8) | cmd.header[1]; // parse command type value
            return true;
        }
        return false;
    }
            
    unsigned int Li1Radio::parseCmdPayload(Li1Cmd & cmd, unsigned int payloadSize, const vector<uint8_t> & serBytes) {
        if (serBytes.size() >= (payloadSize + Li1Radio::checksumLen)) {
            // Check payload checksum
            vector<uint8_t> payloadBytes = vector<uint8_t>(serBytes.begin(), serBytes.begin() + payloadSize);
            vector<uint8_t> headerAndPayload = cmd.rawHeader;
            headerAndPayload.insert(headerAndPayload.end(), payloadBytes.begin(), payloadBytes.end());
            vector<uint8_t> payloadChecksum = vector<uint8_t>(serBytes.begin() + payloadSize, serBytes.begin() + (payloadSize + Li1Radio::checksumLen));    
            
            if (compareChecksum(headerAndPayload, payloadChecksum) == true) {
                cmd.payload = parseAX25Msg(payloadBytes);
            }
            return payloadSize + Li1Radio::checksumLen;
        }
        else { // incomplete command - too short
            return serBytes.size();
        }
    }
            
    vector<uint8_t> Li1Radio::parseAX25Msg(vector<uint8_t> & msgBytes) {
        return vector<uint8_t>(msgBytes.begin()+16, msgBytes.end()-2);
    }
            
    unsigned int Li1Radio::parseCommand(Li1Cmd & cmd, vector<uint8_t> & serBytes) {
        cmd.valid = false; 

        for (unsigned int i = 0; i < (serBytes.size() - (Li1Radio::lenSyncBytes-1)); i++) {
            // Search for sync characters
            if (serBytes[i] == Li1Radio::syncBytes[0] && serBytes[i+1] == Li1Radio::syncBytes[1]) {
                unsigned int msgStart = i;
            
                bool headerFound = false;
                if ((serBytes.size() - msgStart) >= Li1Radio::headerLength) { // entire header received
                    headerFound = parseCmdHeader(cmd, vector<uint8_t>(serBytes.begin() + msgStart, serBytes.end()));
                }
                else { // incomplete command
                    return serBytes.size();
                }

                unsigned int msgEnd = msgStart + Li1Radio::headerLength; // update buffer position
                if (headerFound == true) {
                    // Check for payload
                    if (cmd.type == LI1_RECEIVED_DATA) {
                        unsigned int payloadSize = ((uint16_t)cmd.header[2] << 8) | cmd.header[3]; // reverse payload size bytes 
                        if (payloadSize == 0) { // no payload
                            cmd.valid = true;
                            return msgEnd;
                        }
                        else if (payloadSize == 65535) { // receive error status
                            cmd.valid = true;
                            return msgEnd;
                        }
                        else { // parse payload
                            if (serBytes.size() >= (Li1Radio::headerLength + payloadSize + Li1Radio::checksumLen)) { // entire message received
                                unsigned int payloadEnd = parseCmdPayload(cmd, payloadSize, vector<uint8_t>(serBytes.begin() + msgStart + Li1Radio::headerLength, serBytes.end()));
                                msgEnd += payloadEnd;
                                if (cmd.payload.size() > 0) { // payload found
                                    cmd.valid = true;
                                    return msgEnd;
                                }
                                else {
                                    return msgEnd; // return with invalid cmd flag
                                }
                            }
                            else {
                                return serBytes.size();
                            }
                        }
                    }
                    else { // header only valid command
                        cmd.valid = true;
                        return msgEnd;  
                    }
                }
                else { // invalid header
                    return msgEnd;
                }
            }
        }
                    
        return serBytes.size();
    }
            
    unsigned int Li1Radio::sendMsg(vector<uint8_t> & msgBytes) {
        if (msgBytes.size() > 0) {
            unsigned int numMsgs = ceil(msgBytes.size() / (double)Li1Radio::maxPayload);
            for (unsigned int i = 0; i < numMsgs; i++) {
                // End of message location
                unsigned int msgEnd = 0;
                if (msgBytes.size() > ((i+1)*Li1Radio::maxPayload)) {    
                    msgEnd = (i+1)*Li1Radio::maxPayload;
                }
                else {
                    msgEnd = msgBytes.size();
                }
                
                // Send message
                vector<uint8_t> payload = vector<uint8_t>(msgBytes.begin() + i*Li1Radio::maxPayload, msgBytes.begin() + msgEnd);
                Radio::sendMsg(payload);
                
            }
            return numMsgs;
        }

        return 0;
    }
    
    int Li1Radio::processRxBytes(vector<uint8_t> & newBytes, bool bufferFlag) {
        Li1Cmd cmd;
        
        // Search received bytes for commands
        cmdRxBuffer.insert(cmdRxBuffer.end(), newBytes.begin(), newBytes.end());
        unsigned int msgEnd = 0;
        unsigned int loopCtr = 0;
        while (msgEnd < cmdRxBuffer.size() && loopCtr < 5) {
            msgEnd = parseCommand(cmd, cmdRxBuffer);
        
            if (cmd.valid == true) { // command found
                cmdRxBuffer = vector<uint8_t>(cmdRxBuffer.begin() + msgEnd, cmdRxBuffer.end()); // clear out parsed bytes
                cmdBuffer.push(cmd); // add to parsed command buffer
                
                // Check for received data
                if (cmd.type == LI1_RECEIVED_DATA && cmd.payload.size() > 0) {
                    bufferRxMsg(cmd.payload, bufferFlag);
                }
            }
        }

        if (cmdRxBuffer.size() > config.cmdRxBufferSize) { // buffer too large, clear buffer
            cmdRxBuffer.clear();
        } 
        return 0;
    }
}
